### PR TITLE
API Contract Tests

### PR DESCRIPTION
**Pull request by chartjes**
@mythmon Osmose r?

### CODE REVIEW

**Comment by mythmon** [comment]
When running this, I get an import error for `firefox_puppeteer`. Installing that manually via pip makes the test run. I get really cryptic errors about "NoSectionError: No section: None" until I add `--env dev`. After adding that, the tests pass for me.

File: normandy/conftest.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    )
+    parser.addoption(
+        '--bin',
+        default='/Applications/FirefoxBeta.app/Contents/MacOS/firefox-bin',
+        help='path for Firefox binary'
+    )
+    parser.addoption(
+        "--env",
+        action="store",
+        help="Choose a test environment: stage or prod"
```
Comment (mythmon) [inline normandy/conftest.py]: Adding `required=True` here makes py.test give much better errors when `env` isn't passed.

File: normandy/conftest.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+from firefox_puppeteer import Puppeteer
+from marionette_driver.marionette import Marionette
+
+TIMEOUT = 20
+
+
+def pytest_addoption(parser):
+    parser.addoption(
+        "--admin-url",
```
Comment (mythmon) [inline normandy/conftest.py]: I don't think this option is used anywhere, though I can certainly image it being useful in the future.

Comment (chartjes) [inline normandy/conftest.py]: > I don't think this option is used anywhere, though I can certainly image it being
> useful in the future.

That is a relic from when I was trying to get Marionette tests with the admin UI working.

**Comment by mythmon** [comment]
@Osmose you were tagged for review here, but not notified.

File: normandy/schema-check/test_api.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+
+
+@pytest.fixture
+def conf():
+    config = configparser.ConfigParser()
+    config.read('manifest.ini')
+    return config
+
+
+def test_recipies(conf, env):
```
Comment (rpappalax) [inline normandy/schema-check/test_api.py]: spelling error

File: normandy/schema-check/test_api.py [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+        'postAnswerUrl',
+        'learnMoreUrl',
+        'learnMoreMessage'
+    ]
+
+    # Verify we have at least one response
+    assert len(response) >= 1
+
+    # Take a look at the first response and look for the expected fields
+    record = response[0]
```
Comment (Osmose) [inline normandy/schema-check/test_api.py:61]: The first recipe isn't guaranteed to be a `show-heartbeat` recipe, which means assertions about `show-heartbeat`-specific fields (like the survey fields, default fields, and top-level argument fields) may fail even though the API is returning a valid response.

File: normandy/schema-check/test_api.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+        'message',
+        'thanksMessage',
+        'engagementButtonLabel',
+        'postAnswerUrl',
+        'learnMoreUrl',
+        'learnMoreMessage'
+    ]
+
+    # Verify we have at least one response
+    assert len(response) >= 1
```
Comment (Osmose) [inline normandy/schema-check/test_api.py:58]: Is having no recipes whatsoever considered a failure state? From the app's perspective it's not, and at least from my standpoint having no recipes isn't necessarily wrong, although I don't expect we'll ever really not have at least one recipe in dev, stage, and prod.

File: normandy/schema-check/test_api.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+    expected_action_fields = [
+        'name',
+        'implementation_url',
+        'arguments_schema'
+    ]
+    for field in record:
+        assert field in expected_action_fields
+
+    # Do we have the correct 'arguments_schema' fields?
+    expected_arguments_schema_fields = [
```
Comment (Osmose) [inline normandy/schema-check/test_api.py]: Similar to my comment above, the only hard requirement of the `arguments_schema` field is that it is a valid JSON Schema; not having a `description` field would still be a valid response from the API.

The [jsonschema](https://github.com/Julian/jsonschema/) library has some functions that can help with validating a schema if you're interested in that.

File: normandy/conftest.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+import pytest
+
+TIMEOUT = 20
+
+
+def pytest_addoption(parser):
+    parser.addoption(
+        '--bin',
+        default='/Applications/FirefoxBeta.app/Contents/MacOS/firefox-bin',
+        help='path for Firefox binary',
```
Comment (rpappalax) [inline normandy/conftest.py]: the user should know this is only for 1 OS.  perhaps something like:
help='path to Firefox binary (OSX only)'

**Comment by Osmose** [comment]
@chartjes Do you want to meet in person to talk about this API? Without [api docs](https://bugzilla.mozilla.org/show_bug.cgi?id=1279097), I think that'd be the best way to convey some of my thoughts.

**Pull request closed without merging**